# Comparing `tmp/Palos-0.1.8-py3-none-any.whl.zip` & `tmp/Palos-0.1.9-py3-none-any.whl.zip`

## zipinfo {}

```diff
@@ -1,8 +1,8 @@
-Zip file size: 1018222 bytes, number of entries: 307
+Zip file size: 1018220 bytes, number of entries: 307
 -rw-rw-r--  2.0 unx     2990 b- defN 20-Apr-01 12:33 Genome/OutputGenomeAnnotation.py
 -rw-rw-r--  2.0 unx     9368 b- defN 20-Apr-01 12:33 Genome/__init__.py
 -rw-rw-r--  2.0 unx      934 b- defN 20-Jan-05 09:56 Genome/calculateTotalBEDSpan.py
 -rw-rw-r--  2.0 unx    18798 b- defN 20-Jan-05 09:56 Genome/chromosome_fasta2db.py
 -rw-rw-r--  2.0 unx    16866 b- defN 20-Mar-30 13:06 algorithm/BinarySearchTree.py
 -rw-rw-r--  2.0 unx     3872 b- defN 20-Jan-05 09:56 algorithm/PCA.py
 -rw-rw-r--  2.0 unx     2756 b- defN 20-Jan-05 09:56 algorithm/PartitionGraphIntoCliques.py
@@ -133,15 +133,15 @@
 -rw-rw-r--  2.0 unx     1334 b- defN 20-Apr-11 11:37 ngs/qc/fastq_to_sam_step2.py
 -rw-rw-r--  2.0 unx      522 b- defN 20-Apr-11 11:37 ngs/qc/flagstat.py
 -rw-rw-r--  2.0 unx     2881 b- defN 20-Apr-12 13:14 ngs/qc/get_fastq_name_step1.py
 -rw-rw-r--  2.0 unx     2009 b- defN 20-Apr-11 11:37 ngs/qc/sam_to_fastq_step3.py
 -rw-r--r--  2.0 unx     9274 b- defN 20-May-09 07:43 palos/Genome.py
 -rw-r--r--  2.0 unx    24311 b- defN 20-Apr-16 12:22 palos/MPIwrapper.py
 -rw-r--r--  2.0 unx    30140 b- defN 20-May-18 02:37 palos/ProcessOptions.py
--rw-r--r--  2.0 unx      993 b- defN 20-May-18 06:47 palos/__init__.py
+-rw-r--r--  2.0 unx      993 b- defN 20-May-18 06:55 palos/__init__.py
 -rw-r--r--  2.0 unx    36267 b- defN 20-Apr-14 12:52 palos/gviz_api.py
 -rw-r--r--  2.0 unx    11702 b- defN 20-May-18 02:37 palos/statistics.py
 -rw-r--r--  2.0 unx    40601 b- defN 20-May-12 08:14 palos/utils.py
 -rw-r--r--  2.0 unx    16866 b- defN 20-Apr-14 12:52 palos/algorithm/BinarySearchTree.py
 -rw-r--r--  2.0 unx     3867 b- defN 20-Apr-23 11:35 palos/algorithm/PCA.py
 -rw-r--r--  2.0 unx     2758 b- defN 20-Apr-16 12:22 palos/algorithm/PartitionGraphIntoCliques.py
 -rw-r--r--  2.0 unx    29308 b- defN 20-Apr-23 11:35 palos/algorithm/RBTree.py
@@ -166,19 +166,19 @@
 -rw-r--r--  2.0 unx     3392 b- defN 20-Apr-23 11:35 palos/io/latex.py
 -rw-r--r--  2.0 unx     4331 b- defN 20-Apr-27 11:29 palos/mapper/AbstractDBJob.py
 -rw-r--r--  2.0 unx     4759 b- defN 20-Apr-27 11:29 palos/mapper/AbstractMapper.py
 -rw-r--r--  2.0 unx     2162 b- defN 20-Apr-27 11:29 palos/mapper/AbstractSunsetMapper.py
 -rw-r--r--  2.0 unx     1231 b- defN 20-Apr-27 11:29 palos/mapper/AbstractVCFMapper.py
 -rw-r--r--  2.0 unx       43 b- defN 20-Apr-14 12:52 palos/mapper/__init__.py
 -rw-r--r--  2.0 unx     4193 b- defN 20-May-13 04:41 palos/ngs/AbstractAlignmentAndVCFWorkflow.py
--rw-r--r--  2.0 unx    40234 b- defN 20-May-18 06:37 palos/ngs/AbstractAlignmentWorkflow.py
+-rw-r--r--  2.0 unx    40239 b- defN 20-May-18 06:50 palos/ngs/AbstractAlignmentWorkflow.py
 -rw-r--r--  2.0 unx   157680 b- defN 20-May-18 06:26 palos/ngs/AbstractNGSWorkflow.py
 -rw-r--r--  2.0 unx    65172 b- defN 20-May-15 13:54 palos/ngs/AbstractVCFWorkflow.py
 -rw-r--r--  2.0 unx    63492 b- defN 20-May-15 13:54 palos/ngs/GenericVCFWorkflow.py
--rw-r--r--  2.0 unx    27894 b- defN 20-May-18 06:36 palos/ngs/MapReduceGenomeFileWorkflow.py
+-rw-r--r--  2.0 unx    27899 b- defN 20-May-18 06:50 palos/ngs/MapReduceGenomeFileWorkflow.py
 -rw-r--r--  2.0 unx     7411 b- defN 20-May-11 04:14 palos/ngs/TestMapReduceVCFWorkflow.py
 -rw-r--r--  2.0 unx     5001 b- defN 20-May-12 01:30 palos/ngs/__init__.py
 -rw-r--r--  2.0 unx     4812 b- defN 20-Apr-23 11:35 palos/ngs/io/AlignmentDepthIntervalFile.py
 -rw-r--r--  2.0 unx     3531 b- defN 20-Apr-23 11:35 palos/ngs/io/BamFile.py
 -rw-r--r--  2.0 unx    38342 b- defN 20-May-18 04:24 palos/ngs/io/VCFFile.py
 -rw-r--r--  2.0 unx        0 b- defN 20-Apr-14 12:52 palos/ngs/io/__init__.py
 -rw-r--r--  2.0 unx    10094 b- defN 20-May-18 02:40 palos/pedigree/PlinkPedigreeFile.py
@@ -296,14 +296,14 @@
 -rw-rw-r--  2.0 unx    12307 b- defN 20-Jan-05 09:56 yhio/AbstractMatrixFileWalker.py
 -rw-rw-r--  2.0 unx     3209 b- defN 20-Jan-05 09:56 yhio/FastaFile.py
 -rw-rw-r--  2.0 unx    23045 b- defN 20-Apr-01 12:33 yhio/HDF5MatrixFile.py
 -rw-rw-r--  2.0 unx     7875 b- defN 20-Jan-05 09:56 yhio/MatrixFile.py
 -rw-rw-r--  2.0 unx    29277 b- defN 20-Mar-30 13:06 yhio/YHPyTables.py
 -rw-rw-r--  2.0 unx      223 b- defN 20-Apr-11 11:37 yhio/__init__.py
 -rw-rw-r--  2.0 unx     3391 b- defN 20-Jan-05 09:56 yhio/latex.py
--rw-r--r--  2.0 unx     1142 b- defN 20-May-18 06:48 Palos-0.1.8.dist-info/DESCRIPTION.rst
--rw-r--r--  2.0 unx      862 b- defN 20-May-18 06:48 Palos-0.1.8.dist-info/metadata.json
--rw-r--r--  2.0 unx        6 b- defN 20-May-18 06:48 Palos-0.1.8.dist-info/top_level.txt
--rw-r--r--  2.0 unx       92 b- defN 20-May-18 06:48 Palos-0.1.8.dist-info/WHEEL
--rw-r--r--  2.0 unx     1854 b- defN 20-May-18 06:48 Palos-0.1.8.dist-info/METADATA
--rw-r--r--  2.0 unx    27655 b- defN 20-May-18 06:48 Palos-0.1.8.dist-info/RECORD
-307 files, 3876867 bytes uncompressed, 974990 bytes compressed:  74.9%
+-rw-r--r--  2.0 unx     1142 b- defN 20-May-18 06:56 Palos-0.1.9.dist-info/DESCRIPTION.rst
+-rw-r--r--  2.0 unx      862 b- defN 20-May-18 06:56 Palos-0.1.9.dist-info/metadata.json
+-rw-r--r--  2.0 unx        6 b- defN 20-May-18 06:56 Palos-0.1.9.dist-info/top_level.txt
+-rw-r--r--  2.0 unx       92 b- defN 20-May-18 06:56 Palos-0.1.9.dist-info/WHEEL
+-rw-r--r--  2.0 unx     1854 b- defN 20-May-18 06:56 Palos-0.1.9.dist-info/METADATA
+-rw-r--r--  2.0 unx    27655 b- defN 20-May-18 06:56 Palos-0.1.9.dist-info/RECORD
+307 files, 3876877 bytes uncompressed, 974988 bytes compressed:  74.9%
```

## zipnote {}

```diff
@@ -897,26 +897,26 @@
 
 Filename: yhio/__init__.py
 Comment: 
 
 Filename: yhio/latex.py
 Comment: 
 
-Filename: Palos-0.1.8.dist-info/DESCRIPTION.rst
+Filename: Palos-0.1.9.dist-info/DESCRIPTION.rst
 Comment: 
 
-Filename: Palos-0.1.8.dist-info/metadata.json
+Filename: Palos-0.1.9.dist-info/metadata.json
 Comment: 
 
-Filename: Palos-0.1.8.dist-info/top_level.txt
+Filename: Palos-0.1.9.dist-info/top_level.txt
 Comment: 
 
-Filename: Palos-0.1.8.dist-info/WHEEL
+Filename: Palos-0.1.9.dist-info/WHEEL
 Comment: 
 
-Filename: Palos-0.1.8.dist-info/METADATA
+Filename: Palos-0.1.9.dist-info/METADATA
 Comment: 
 
-Filename: Palos-0.1.8.dist-info/RECORD
+Filename: Palos-0.1.9.dist-info/RECORD
 Comment: 
 
 Zip file comment:
```

## palos/__init__.py

```diff
@@ -1,13 +1,13 @@
 """
 Palos is a Python3 module developed and used by the yfish group, http://www.yfish.org/.
 It contains code related to bioinformatics projects focusing on next-generation sequencing data, 
 population genetics, genome-wide association studies, pedigree genetics, etc.
 """
-version='0.1.8'
+version='0.1.9'
 from . ProcessOptions import ProcessOptions, generate_program_doc, process_options, \
 	process_function_arguments, turn_option_default_dict2argument_default_dict
 from . utils import PassingData, PassingDataList, dict_map, importNumericArray, \
 	figureOutDelimiter, get_gene_symbol2gene_id_set, \
 	FigureOutTaxID, getColName2IndexFromHeader, getListOutOfStr, runLocalCommand, \
 	getGeneIDSetGivenAccVer, calGreatCircleDistance, openGzipFile
 from . Genome import GeneModel
```

## palos/ngs/AbstractAlignmentWorkflow.py

```diff
@@ -872,15 +872,15 @@
             # 2012.8.2 if maxContigID/minContigID is not well defined.
             #  restrictContigDictionry won't do anything.
             chr2IntervalDataLs = self.restrictContigDictionry(
                 dc=chr2IntervalDataLs,
                 maxContigID=self.maxContigID, minContigID=self.minContigID)
         else:
             logging.warn(f"self.needSplitChrIntervalData="\
-                f"{needSplitChrIntervalData}, set chr2IntervalDataLs=None.")
+                f"{self.needSplitChrIntervalData}, set chr2IntervalDataLs=None.")
             chr2IntervalDataLs = None
 
         alignmentLs = self.getAlignments()
         alignmentDataLs = self.registerAlignmentAndItsIndexFile(
             alignmentLs=alignmentLs, data_dir=self.data_dir)
         self.alignmentLs = alignmentLs
         self.alignmentDataLs = alignmentDataLs
```

## palos/ngs/MapReduceGenomeFileWorkflow.py

```diff
@@ -611,15 +611,15 @@
             # defined in __init__().
             chr2IntervalDataLs = self.getChr2IntervalDataLsBySplitChrSize(
                 chr2size=self.chr2size,
                 intervalSize=self.intervalSize,
                 intervalOverlapSize=self.intervalOverlapSize)
         else:
             logging.warn(f"self.needSplitChrIntervalData="\
-                f"{needSplitChrIntervalData}, set chr2IntervalDataLs=None.")
+                f"{self.needSplitChrIntervalData}, set chr2IntervalDataLs=None.")
             chr2IntervalDataLs = None
         inputData = None
         firstInputJobData = None
         if getattr(self, 'inputDir', None):
             inputData = self.registerFilesOfInputDir(
                 inputDir=self.inputDir,
                 input_site_handler=self.input_site_handler, \
```

## Comparing `Palos-0.1.8.dist-info/DESCRIPTION.rst` & `Palos-0.1.9.dist-info/DESCRIPTION.rst`

 * *Files identical despite different names*

## Comparing `Palos-0.1.8.dist-info/metadata.json` & `Palos-0.1.9.dist-info/metadata.json`

 * *Files 12% similar despite different names*

### Pretty-printed

 * *Similarity: 0.9583333333333334%*

 * *Differences: {"'version'": "'0.1.9'"}*

```diff
@@ -39,9 +39,9 @@
                 "numpy",
                 "pegaflow",
                 "sqlalchemy"
             ]
         }
     ],
     "summary": "Misc Python modules developed and used by the yfish group",
-    "version": "0.1.8"
+    "version": "0.1.9"
 }
```

## Comparing `Palos-0.1.8.dist-info/METADATA` & `Palos-0.1.9.dist-info/METADATA`

 * *Files 0% similar despite different names*

```diff
@@ -1,10 +1,10 @@
 Metadata-Version: 2.0
 Name: Palos
-Version: 0.1.8
+Version: 0.1.9
 Summary: Misc Python modules developed and used by the yfish group
 Home-page: https://github.com/polyactis/pymodule
 Author: Yu S. Huang
 Author-email: polyactis@gmail.com
 License: Apache2
 Platform: UNKNOWN
 Classifier: Intended Audience :: Developers
```

## Comparing `Palos-0.1.8.dist-info/RECORD` & `Palos-0.1.9.dist-info/RECORD`

 * *Files 1% similar despite different names*

```diff
@@ -1,17 +1,17 @@
 Genome/OutputGenomeAnnotation.py,sha256=yx1BBAS1la9IDYDuSB9eGNJUmy3Ybsb1gHGBAPyGzLM,2990
 Genome/__init__.py,sha256=HFwYp4UxkSny7swTd_jBhtBtFRd6oeqMhu-tt1Jng2Q,9368
 Genome/calculateTotalBEDSpan.py,sha256=Rt3ydLDKosfci-2ONazJryVT1JMK2Ab9WB1SM6sie-k,934
 Genome/chromosome_fasta2db.py,sha256=iaaMJDXfnVLLy51JLCjPxbFaCVCAbIAbqX6-DSn0CX0,18798
-Palos-0.1.8.dist-info/DESCRIPTION.rst,sha256=DtK3Hcr7NHf49duCkcUANLd17gN_QCl3JQNAI-4mzpQ,1142
-Palos-0.1.8.dist-info/METADATA,sha256=tYgBek-52lSFeTBmabArj_etZz7buvOZbhDNx9x6ZyE,1854
-Palos-0.1.8.dist-info/RECORD,,
-Palos-0.1.8.dist-info/WHEEL,sha256=8Lm45v9gcYRm70DrgFGVe4WsUtUMi1_0Tso1hqPGMjA,92
-Palos-0.1.8.dist-info/metadata.json,sha256=ICIa2vAAN-VwkU7Wm0aLN-LMdEFPOdHAg5i4a7DOGqk,862
-Palos-0.1.8.dist-info/top_level.txt,sha256=rSB8QUNf_REJVGQvz5092SxgiaIFskDspv6zmAo-xAk,6
+Palos-0.1.9.dist-info/DESCRIPTION.rst,sha256=DtK3Hcr7NHf49duCkcUANLd17gN_QCl3JQNAI-4mzpQ,1142
+Palos-0.1.9.dist-info/METADATA,sha256=gi3DWRF17D31CGZdzS-HB9ZeaeN4-TVCn75WRzufhjM,1854
+Palos-0.1.9.dist-info/RECORD,,
+Palos-0.1.9.dist-info/WHEEL,sha256=8Lm45v9gcYRm70DrgFGVe4WsUtUMi1_0Tso1hqPGMjA,92
+Palos-0.1.9.dist-info/metadata.json,sha256=FiJ9TVnAMSLavLEbeMwfasDEQKZM513dWwq0d88rNg8,862
+Palos-0.1.9.dist-info/top_level.txt,sha256=rSB8QUNf_REJVGQvz5092SxgiaIFskDspv6zmAo-xAk,6
 algorithm/BinarySearchTree.py,sha256=oM-2MVFC1ZW6ebNCJnOzP4_63EtpVd-ec13CvdRalL4,16866
 algorithm/PCA.py,sha256=--ANTa656Xs3qJN9whC5kqm-vl69xbtBoDDkWmWz1oU,3872
 algorithm/PartitionGraphIntoCliques.py,sha256=9KkjEk7w-ZLTbmjflWv1IPq65mawYRaHegoEXQFdEXE,2756
 algorithm/RBTree.py,sha256=DdWDefYjKVbShi_OVeLTAGNb_3-c2_yZ-nMrsrhlrwE,29430
 algorithm/__init__.py,sha256=XG8Q5f1wWX3Z9a4N23xJlb-2IoUnG4Tko6Uv4WRbUtY,12936
 algorithm/clique.py,sha256=WCtUAsJ-b73agyRqQu9q_61wOisn4pMWVtTEmqLZBik,2580
 algorithm/graph.py,sha256=6w4kqkgyICj8MzUlQmZgJRhY6SJL-oaELwWEywLk8bw,7242
@@ -138,15 +138,15 @@
 ngs/qc/fastq_to_sam_step2.py,sha256=-Kl64bVqwnvFxGD-AdEEz_ZHMJC4WRR4EXWpULOJJzY,1334
 ngs/qc/flagstat.py,sha256=yTagumzbHKCKks89lsjFVzTbbG_NqA8JLKzt4c1k_UI,522
 ngs/qc/get_fastq_name_step1.py,sha256=yVT45dO8qpEMjGDC6KdIR03vB1sX9gIW8GjbPZV_uEY,2881
 ngs/qc/sam_to_fastq_step3.py,sha256=gr3z_6Kf2zvtIKJdkI8sX7xyuNCDsxSTbBFeHTmpqxk,2009
 palos/Genome.py,sha256=K0GmUfg2msm22SW1m1rEusn-heHUxE9E32WUpyN-z1U,9274
 palos/MPIwrapper.py,sha256=U0-2A8ICC98SqOuPL8EEJovrTjFemV9bYHM3H41iLqE,24311
 palos/ProcessOptions.py,sha256=X4Y9IoNuG4SelaeqLZymSBOU2A1IAHJYY34zKjY_wMk,30140
-palos/__init__.py,sha256=3Lo-o7PNTwktEcBs9IE1kAZAijIcw6IctazgifrylPk,993
+palos/__init__.py,sha256=lo18ThNFJNrV_n9GiRRj_3rPlu9jLRBu9veRJsJOT5I,993
 palos/gviz_api.py,sha256=Ayn6coBoFq9Np5XG2-nmNtl3xr5k7yULVs1rZjomm08,36267
 palos/statistics.py,sha256=3VSrUv_hpGD9y374hiX5Sccxhnq7938vdweYAW-UigE,11702
 palos/utils.py,sha256=pCOtbcoaWWZanE0n2-Bs3I90XAJ-GTAvr1qitWSPfh8,40601
 palos/algorithm/BinarySearchTree.py,sha256=oM-2MVFC1ZW6ebNCJnOzP4_63EtpVd-ec13CvdRalL4,16866
 palos/algorithm/PCA.py,sha256=7kMk_z7eGZvly18Y4xRQ87GKM3vsu1Xl46iCze3zl4w,3867
 palos/algorithm/PartitionGraphIntoCliques.py,sha256=nVxCIbhobzZ603NGXZeKzwY0v4iteSmbW0SDoZzW8s0,2758
 palos/algorithm/RBTree.py,sha256=ORlmSCIfwVB_yPxwK2ZHqLuZFyWYRERd7N5Z2oQ0UtQ,29308
@@ -171,19 +171,19 @@
 palos/io/latex.py,sha256=RreLst1E0ExqhROh3WscAFHJ1LCQIxN5Ep5ey4ZqxmM,3392
 palos/mapper/AbstractDBJob.py,sha256=skWlErWSqvOCeEr2YGCdhoHxWfw5_MSVfSnqGd6GF3U,4331
 palos/mapper/AbstractMapper.py,sha256=wuWDxA_Zy5jfiV6EUYD00b7A-0Y2TT145du9LeBcz2A,4759
 palos/mapper/AbstractSunsetMapper.py,sha256=eRKLISayrOA-rp4JvlDpGSgFtm8bRdvHGkfkFijI-Cc,2162
 palos/mapper/AbstractVCFMapper.py,sha256=tRTaIYGVDYFb3VdejicdfMjk9X19zzgsLLY2LSbj-bg,1231
 palos/mapper/__init__.py,sha256=Max0ET15cwoOKoyslo28yKJdgw3lTavM9VLdsUOzrLw,43
 palos/ngs/AbstractAlignmentAndVCFWorkflow.py,sha256=t5D-QNl1oXQm4KN4TqGf8P5szzm4T0EhZ8YM7qviUBA,4193
-palos/ngs/AbstractAlignmentWorkflow.py,sha256=9LIpavcaaxNnPh4RqmqXCQMqyKowoO8gr6TwKDdpGi0,40234
+palos/ngs/AbstractAlignmentWorkflow.py,sha256=tbWbH9uH6bcpUe3ZY-_QPzYkUXMaVDgU-mgnnUn_ZQk,40239
 palos/ngs/AbstractNGSWorkflow.py,sha256=ELklhn90WPQIXdyOIfobu5eMuxy359x4iHECWN2VhNA,157680
 palos/ngs/AbstractVCFWorkflow.py,sha256=6KcnGaCvXnZpKrqJe6tJUK0QY8WDS7ZRGVmZ95zrTCk,65172
 palos/ngs/GenericVCFWorkflow.py,sha256=X7sI92qFSvGDh8Un_tgfQvlWxQfxf6kAkhdRNKwBsok,63492
-palos/ngs/MapReduceGenomeFileWorkflow.py,sha256=Y55IGo7eRpb4LsJN_veuYMtlgKI-EfTnxBI2T5j9-UI,27894
+palos/ngs/MapReduceGenomeFileWorkflow.py,sha256=PgZOYNsuX-jOI3bMyUpIjSWJid3cb7Q8Z1k14KI4i5A,27899
 palos/ngs/TestMapReduceVCFWorkflow.py,sha256=j5wsgeswXtW1I7CmU6DgTbgz5wlNFwkLZf_C5iCCX6c,7411
 palos/ngs/__init__.py,sha256=-dZDjHpsaGCG9taFC3r6bHZd2gk97YvXtPcfWxNb38M,5001
 palos/ngs/io/AlignmentDepthIntervalFile.py,sha256=lBgRgD5nZC10KOT4g9sNR76NPO6DMChOaor2NM8ChmU,4812
 palos/ngs/io/BamFile.py,sha256=JV_RDE7iK1HB7m99cfkO0MjJuJdozsMZ9thqMk7z2oQ,3531
 palos/ngs/io/VCFFile.py,sha256=-5wl8EJS8z7lupLSpTsboIeKQXTl6s0nt2ZS98_JKMI,38342
 palos/ngs/io/__init__.py,sha256=47DEQpj8HBSa-_TImW-5JCeuQeRkm5NMpJWZG3hSuFU,0
 palos/pedigree/PlinkPedigreeFile.py,sha256=eqodelK3SnnsiQ-RtuqismiQqS3hmGh7LIN_TiR5zu4,10094
```

